File: src/GlobalServerEvent.ts
import NetServerEvent from "./ServerEvent";
import NetGlobalEvent from "./GlobalEvent";
import { getGlobalRemote, IS_CLIENT, isLuaTable } from "./internal";
const Players = game.GetService("Players");

export interface IMessage {
	data: Array<unknown>;
	targetId?: number;
	targetIds?: Array<number>;
}

function isMessage(value: unknown): value is IMessage {
	if (isLuaTable(value)) {
		const hasData = value.has("data");
		return !value.isEmpty() && (hasData && typeOf(value.get("data")) === "table");
	} else {
		return false;
	}
}

/**
 * Similar to a ServerEvent, but works across all servers.
 */
export default class NetGlobalServerEvent implements INetXServerEvent {
	private readonly instance: NetServerEvent;
	private readonly event: NetGlobalEvent;

	constructor(name: string) {
		this.instance = new NetServerEvent(getGlobalRemote(name));
		this.event = new NetGlobalEvent(name);
		assert(!IS_CLIENT, "Cannot create a Net.GlobalServerEvent on the Client!");

		this.event.Connect(message => {
			if (isMessage(message)) {
				this.recievedMessage(message);
			} else {
				warn(`[rbx-net] Recieved malformed message for GlobalServerEvent: ${name}`);
			}
		});
	}

	private getPlayersMatchingId(matching: Array<number> | number) {
		if (typeIs(matching, "number")) {
			return Players.GetPlayerByUserId(matching);
		} else {
			const players = new Array<Player>();
			for (const id of matching) {
				const player = Players.GetPlayerByUserId(id);
				if (player) {
					players.push(player);
				}
			}

			return players;
		}
	}

	private recievedMessage(message: IMessage) {
		if (message.targetIds) {
			const players = this.getPlayersMatchingId(message.targetIds);
			if (players) {
				this.instance.SendToPlayers(players as Array<Player>, ...message.data);
			}
		} else if (message.targetId) {
			const player = this.getPlayersMatchingId(message.targetId);
			if (player) {
				this.instance.SendToPlayer(player as Player, ...message.data);
			}
		} else {
			this.instance.SendToAllPlayers(...message.data);
		}
	}

	/**
	 * SEnds an event to all servers in the game
	 * @param args The args of the message
	 */
	public SendToAllServers<T extends Array<unknown>>(...args: T) {
		this.event.SendToAllServers({ data: [...args] });
	}

	/**
	 * Sends an event to the specified server
	 * @param jobId The game.JobId of the target server
	 * @param args The args of the message
	 */
	public SendToServer<T extends Array<unknown>>(jobId: string, ...args: T) {
		this.event.SendToServer(jobId, { data: [...args] });
	}

	/**
	 * Sends an event to the specified player (if they're in any of the game's servers)
	 * @param userId The userId of the target player
	 * @param args The args
	 */
	public SendToPlayer<T extends Array<unknown>>(userId: number, ...args: T) {
		const player = Players.GetPlayerByUserId(userId);
		// If the player exists in this instance, just send it straight to them.
		if (player) {
			this.instance.SendToPlayer(player, ...args);
		} else {
			this.event.SendToAllServers({ data: [...args], targetId: userId });
		}
	}

	/**
	 * Sends an event to all the specified players (if they're in any of the game's servers)
	 * @param userIds The list of user ids to send this message to
	 * @param args The args of the message
	 */
	public SendToPlayers<T extends Array<unknown>>(userIds: Array<number>, ...args: T) {
		// Check to see if any of these users are in this server first, and handle accordingly.
		for (const targetId of userIds) {
			const player = Players.GetPlayerByUserId(targetId);
			if (player) {
				this.instance.SendToPlayer(player, ...args);
				userIds.remove(targetId);
			}
		}

		if (userIds.length > 0) {
			this.event.SendToAllServers({ data: [...args], targetIds: userIds });
		}
	}
}

File: GlobalServerEvent.lua

local TS = _G[script]
local NetServerEvent = TS.import(script, script.Parent, "ServerEvent").default
local _0 = TS.import(script, script.Parent, "GlobalEvent")
local NetGlobalEvent = _0.default
local isSubscriptionMessage = _0.isSubscriptionMessage
local _1 = TS.import(script, script.Parent, "internal")
local getGlobalRemote = _1.getGlobalRemote
local IS_CLIENT = _1.IS_CLIENT
local isLuaTable = _1.isLuaTable
local Players = game:GetService("Players")
local function isTargetedSubscriptionMessage(value)
	if isSubscriptionMessage(value) then
		if isLuaTable(value.Data) then
			return value.Data.InnerData ~= nil
		end
	end
	return false
end
--[[
	*
	* Similar to a ServerEvent, but works across all servers.
]]
local NetGlobalServerEvent
do
	NetGlobalServerEvent = setmetatable({}, {
		__tostring = function()
			return "NetGlobalServerEvent"
		end,
	})
	NetGlobalServerEvent.__index = NetGlobalServerEvent
	function NetGlobalServerEvent.new(...)
		local self = setmetatable({}, NetGlobalServerEvent)
		self:constructor(...)
		return self
	end
	function NetGlobalServerEvent:constructor(name)
		self.instance = NetServerEvent.new(getGlobalRemote(name))
		self.event = NetGlobalEvent.new(name)
		assert(not IS_CLIENT, "Cannot create a Net.GlobalServerEvent on the Client!")
		self.eventHandler = self.event:Connect(function(message)
			if isTargetedSubscriptionMessage(message) then
				self:recievedMessage(message.Data)
			else
				warn("[rbx-net] Recieved malformed message for GlobalServerEvent: " .. name)
			end
		end)
	end
	function NetGlobalServerEvent:getPlayersMatchingId(matching)
		local _2 = matching
		if type(_2) == "number" then
			return Players:GetPlayerByUserId(matching)
		else
			local players = {}
			for _, id in ipairs(matching) do
				local player = Players:GetPlayerByUserId(id)
				if player then
					local _3 = players
					local _4 = player
					
					_3[#_3 + 1] = _4
					
				end
			end
			return players
		end
	end
	function NetGlobalServerEvent:recievedMessage(message)
		local _2 = message.TargetId
		if message.TargetIds then
			local players = self:getPlayersMatchingId(message.TargetIds)
			if players then
				self.instance:SendToPlayers(players, unpack(message.InnerData))
			end
		elseif _2 ~= 0 and _2 == _2 and _2 then
			local player = self:getPlayersMatchingId(message.TargetId)
			if player then
				self.instance:SendToPlayer(player, unpack(message.InnerData))
			end
		else
			self.instance:SendToAllPlayers(unpack(message.InnerData))
		end
	end
	function NetGlobalServerEvent:Disconnect()
		self.eventHandler:Disconnect()
	end
	function NetGlobalServerEvent:SendToAllServers(...)
		local args = { ... }
		local _2 = self.event
		local _3 = {}
		local _4 = "data"
		local _5 = {}
		local _6 = #_5
		for _7, _8 in ipairs(args) do
			_5[_6 + _7] = _8
		end
		_3[_4] = _5
		_2:SendToAllServers(_3)
	end
	function NetGlobalServerEvent:SendToServer(jobId, ...)
		local args = { ... }
		local _2 = self.event
		local _3 = {}
		local _4 = "data"
		local _5 = {}
		local _6 = #_5
		for _7, _8 in ipairs(args) do
			_5[_6 + _7] = _8
		end
		_3[_4] = _5
		_2:SendToServer(jobId, _3)
	end
	function NetGlobalServerEvent:SendToPlayer(userId, ...)
		local args = { ... }
		local player = Players:GetPlayerByUserId(userId)
		
		if player then
			self.instance:SendToPlayer(player, unpack(args))
		else
			local _2 = self.event
			local _3 = {}
			local _4 = "data"
			local _5 = {}
			local _6 = #_5
			for _7, _8 in ipairs(args) do
				_5[_6 + _7] = _8
			end
			_3[_4] = _5
			_3.targetId = userId
			_2:SendToAllServers(_3)
		end
	end
	function NetGlobalServerEvent:SendToPlayers(userIds, ...)
		local args = { ... }
		
		for _, targetId in ipairs(userIds) do
			local player = Players:GetPlayerByUserId(targetId)
			if player then
				self.instance:SendToPlayer(player, unpack(args))
				local _2 = userIds
				local _3 = targetId
				table.remove(_2, _3 + 1)
			end
		end
		if #userIds > 0 then
			local _2 = self.event
			local _3 = {}
			local _4 = "data"
			local _5 = {}
			local _6 = #_5
			for _7, _8 in ipairs(args) do
				_5[_6 + _7] = _8
			end
			_3[_4] = _5
			_3.targetIds = userIds
			_2:SendToAllServers(_3)
		end
	end
end
return {
	default = NetGlobalServerEvent,
}
